### ISSUE TITLE
external dependencies

### ISSUE THREAD

**Issue by nrdxp** [ekala-project/atom#38]
Currently, external dependencies passed to `compose` through its `extern` argument end up being accessible to modules via the top-level `atom` scope. For example, a dependency `A` would be available at `atom.A` if passed to the `extern` arg as `{ extern.A = expressionToGenerate A; }`

However, since this collides with the top-level module of an atom, I wonder if it might not be better to keep external deps to their own scope before we finalize the first iteration of the API. So maybe just a top-level `extern` available everywhere, or perhaps `atom.extern`, however that is fairly verbose.

There is the possibility of just dropping dependencies into the top-level scope as some other languages do, but that risks potential confusion for users who don't know where these things are coming from. Of course, if we get LSP tooling to help out, it may be a non-issue.

Open to suggestions or alternatives, as I'm not sure what the best API is yet, at this time.

**Comment by whs-dot-hk** [comment]
What would passing local repo source into extern looks like? Thanks

**Comment by nrdxp** [comment]
technical `extern` is a low-level API that shouldn't be used most of the time. However, the interface with `eka` cli isn't yet complete, so for now you can use it however you wish. Just keep in mind its not the intended user facing API. The question here is more concerning how users access those external inputs from within a module. Currently it's:
```nix
{
  foo = atom.my-external-dep;
}
```

Here I'm asking if maybe it should be:
```nix
{
  foo = extern.my-external-dep; # extern scope currently doesn't exist
}
# or
{
  foo = atom.extern.my-external-dep; # this technically already exists as an escape hatch when there is a collision. Perhaps we just leave it that way?
}
# or maybe even just
{
  foo = my-external-dep; # dropped into top-level
}
```
Or perhaps something else I'm not thinking of

In terms of how to pass in external source, there is no type check of any kind so you can technically pass whatever you want into extern, including a patch or the results of a network fetch, etc. Once the high-level API is complete the inputs will technically be passed in from the lock file to be able to cleanly manage inputs with the CLI, and you won't ever really have to mess with `extern` yourself.

**Comment by whs-dot-hk** [comment]
```nix
{
  foo = atom.my-external-dep;
}
```
make sense

**Comment by LiGoldragon** [comment]
Hi everyone, I love atom, and am now integrating it into my project.
I'm of the opinion to *use* rusty syntax here.
Rust is the new benchmark now...

```nix
{
  myNixOSSystem = use.nixos-lib.eval-config-minimal { ... };
}
```

**Comment by nrdxp** [comment]
> Hi everyone, I love atom, and am now integrating it into my project. I'm of the opinion to _use_ rusty syntax here. Rust is the new benchmark now...
> 
> ```nix
> {
>   myNixOSSystem = use.nixos-lib.eval-config-minimal { ... };
> }
> ```

oh hey, that's actually not a bad idea. Didn't think of that one.

**Comment by ic4-y** [comment]
Yes, I think the *use* syntax suggestion is quite nice...

**Comment by whs-dot-hk** [comment]
Maybe a override in toml file, able to change the name to `use`, `import`, etc

**Comment by nrdxp** [comment]
> Maybe a override in toml file, able to change the name to `use`, `import`, etc

I'm not sure the added complexity is warranted, and it also would make atom code in the wild potentially more difficult to grok.

**Comment by nrdxp** [comment]
closed by #41
